### PR TITLE
Add new version of the ContentionStart event

### PR DESCRIPTION
**Pull request by gleocadie**
The goal of this PR is to enrich the information provided in the ContentionStart event.

Today, we have the callstack (attached to the ContentionStart event), the contention duration (with the ContentionEnd event).

By adding the ObjectID of the lock object and the ThreadID of the thread that currently holds the lock, we will be able to build the view `which thread is blocking other threads on which lock object`.

### CODE REVIEW

File: src/coreclr/vm/ClrEtwAll.man
Comment (davmason) [inline src/coreclr/vm/ClrEtwAll.man:1748]: `win:Pointer` would be a better data type for LockObjectID

File: src/coreclr/vm/syncblk.cpp
Comment (davmason) [inline src/coreclr/vm/syncblk.cpp:2557]: `ObjectID` and `ThreadID` are ICorProfiler concepts, we shouldn't use them here. A cast shouldn't be necessary at all, if the compiler complains then the cast should be using the data types defined in clretwall.man, i.e. `(UINT64)`.

**Review by davmason — CHANGES_REQUESTED** [review]

File: src/coreclr/vm/ClrEtwAll.man
Comment (gleocadie) [reply src/coreclr/vm/ClrEtwAll.man]: done https://github.com/dotnet/runtime/pull/72627/commits/5fb79fdff0d498b6e456d5a85bc9c9d97d5fb821

File: src/coreclr/vm/syncblk.cpp
Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp]: done https://github.com/dotnet/runtime/pull/72627/commits/5fb79fdff0d498b6e456d5a85bc9c9d97d5fb821

File: src/coreclr/vm/ClrEtwAll.man
Comment (gleocadie) [inline src/coreclr/vm/ClrEtwAll.man:1745]: I wonder if putting the LockId and OwnerId wouldn't be better on the Stop event, WDYT?

Comment (davmason) [reply src/coreclr/vm/ClrEtwAll.man:1745]: I think it makes more sense on the start event, since you want to know what was holding it that triggered the contention

**Review by davmason — APPROVED** [review]
The change looks good to me, but @kouvel should take a look before we merge. He is out of the office for a few more days though

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [inline src/coreclr/vm/syncblk.cpp:2557]: Rather than writing out `m_HoldingThread` which is the address of the `Thread` object in memory, can we change this to `m_HoldingThread->GetOSThreadId64()`?  This will allow a profiler to display what the thread that currently holds the lock is doing.  With the pointer value, you would need to provide another event that correlates between `Thread*` and the actual OS thread ID.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2557]: Another potential issue here is that the value of `obj` can change during GC relocation.  Is the goal to also watch the GC survival and movement events, or do you want to try and have a more durable ID?  I'm wondering if `AwareLock` should also have a set of creation/destruction events so that you can see the full lifetime of the lock.

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2557]: @brianrob good point. I'll make the change to replace the pointer by the OS Thread id.

> Another potential issue here is that the value of obj can change during GC relocation.

Does it mean that the address is changed also ?
The goal, at least in this PR, is not to watch the lock object lifetime, but more having an ID to do some correlation.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2557]: > Does it mean that the address is changed also ?

Yes the object address also changes when it's moved. Could pass in `this` instead to include the `AwareLock` pointer as just a unique identifier for the lock.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2557]: Using `this` is a good idea.  Worth calling out though that it is possible to destroy and then create a new lock at the same address.  To make sure that you know about this, you may wish to write an event in the `AwareLock` ctor so that you know when the aware lock is created.  This type of pattern is other places such as module and type loads.  We use the address of the module or type in memory, but handle questions of lifetime using load/unload events.

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2557]: Thanks for the info.
So if I understand correctly, in the `AwareLock` ctor I emit an event (we know the lock is created) and I will have to emit an event in the dtor to know when the lock is deleted, right ?

I'll do the change taking into account your comment.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2557]: To avoid emitting too many events, I think you can safely emit just the creation event.  If you see two creation events at the same address, you know that it's a new lock.

File: src/coreclr/vm/ClrEtwAll.man
Comment (brianrob) [inline src/coreclr/vm/ClrEtwAll.man:1749]: Recommend calling this `LockID`, and then match with the same name in the `AwareLockCreated` event.

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [inline src/coreclr/vm/syncblk.cpp:2557]: The two new parameters are LockID and OwningThreadID, so the first new parameter should be `this`, and the second should be `pCurThread->GetOSThreadId64()`.

**Review by brianrob — COMMENTED** [review]
A couple of fixups, but overall coming together.

File: src/coreclr/vm/syncblk.cpp
Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2557]: Oh yeah you are right... I forgot this one. I was focused on the AwareLock event.

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2557]: I think it should be `m_HoldingThread` instead of `pCurThread`, because we want the thread that currently holds the lock

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2557]: Yes, you're right.

File: src/coreclr/vm/ClrEtwAll.man
Comment (kouvel) [inline src/coreclr/vm/ClrEtwAll.man:214]: Maybe a new opcode is not necessary, the `win:Start` opcode can be used instead

File: src/coreclr/vm/ClrEtwAll.man
Comment (kouvel) [inline src/coreclr/vm/ClrEtwAll.man:3688]: The name `AwareLock` is specific to the current implementation. The implementation may be changed, or another implementation may be added in the future (eg. https://github.com/dotnet/runtime/issues/34812) and may want to reuse the same events, maybe just `LockCreated` would be more generic

Comment (brianrob) [reply src/coreclr/vm/ClrEtwAll.man:3688]: Agreed, this is a good idea.

File: src/coreclr/vm/ClrEtwAll.man
Comment (brianrob) [reply src/coreclr/vm/ClrEtwAll.man:214]: I think win:Info is a good choice here since we are not starting an action, and there won't be a corresponding stop.

**Comment by tommcdon** [comment]
@kouvel @brianrob @davmason can this PR be merged?

**Comment by brianrob** [comment]
Not yet.  This is waiting for the most recent round of code review feedback to be addressed.

**Comment by tommcdon** [comment]
Thanks for moving the milestone to 8.0 @kouvel.  Given that this PR is still under iteration moving to 8.0 is correct.

**Review by brianrob — APPROVED** [review]
LGTM.  @kouvel, what do you think?

**Comment by brianrob** [comment]
@davmason, it looks like the test failures are related to the new `LockCreated` event, which is triggering an assert.  I suspect this is because the assumptions made in EventPipe code are not met at this call site.  This will probably require some guidance to determine the right fix, though I'm not familiar enough with how this code has evolved to know exactly what to recommend - possibly moving where the event is emitted.

**Comment by davmason** [comment]
> @davmason, it looks like the test failures are related to the new `LockCreated` event, which is triggering an assert. I suspect this is because the assumptions made in EventPipe code are not met at this call site. This will probably require some guidance to determine the right fix, though I'm not familiar enough with how this code has evolved to know exactly what to recommend - possibly moving where the event is emitted.

Looking at it, it should be possible to declare that it is legal to take them in that order and everything should work. Our crst code tries to make everything illegal unless specifically whitelisted.

@gleocadie can you do the following to update it:
- Rebase your changes against the latest in master since there are more recent changes to crsttypes.def than your branch contains
- Add `SyncBlockCache` to the `AcquiredAfter` list for `Crst CodeVersioning` here: https://github.com/dotnet/runtime/blob/40749f6e77f098ca0fae4d88ffd802f49b58c404/src/coreclr/inc/CrstTypes.def#L396-L400
- Regenerate crsttypes.h using the tool from here: https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/CrstTypeTool.cs
  - I find it easiest to create a new project with `dotnet new console`, add the code from `crsttypetool.cs` to `program.cs` and then copy `CrstTypes.def` over to that folder, then run with `dotnet run`. It will generate `crsttypes.h` and you can copy that over in to your enlistment, then check it in.

**Comment by kouvel** [comment]
There are two failures, for the second failure it looks like the place where the LockCreated event is emitted needs to change, which would also make the above lock ordering unnecessary. I also found a separate minor issue. I'll go ahead and post my fixes.

**Comment by kouvel** [comment]
Changes in the latest commit:
- Moved firing the `LockCreated` event to outside the FAULT_FORBID region and outside the `SyncBlockCache` lock
- `m_HoldingThread` may be null. Also, it can't be dereferenced since after reading it the holding thread may release the lock, exit, and the `Thread` object may be deleted. Changed to pass in the `Thread` pointer in `m_HoldingThread` into the event and changed the event to take a pointer for the thread ID. There are already `ThreadCreating` and `ThreadRunning` events that include the native `Thread` pointer.
- Undid some code reordering in `AwareLock::EnterEpilogHelper` since the managed object ref is not used for the event anymore. Also some miscellaneous changes.

**Review by kouvel — APPROVED** [review]

**Comment by kouvel** [comment]
@brianrob and @davmason, if the latest commit looks ok to you, I think we can merge

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [inline src/coreclr/vm/syncblk.cpp:2246]: Does this guarantee that we only fire this event when a lock is created?  There are other paths that create syncblocks, right?

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [inline src/coreclr/vm/syncblk.cpp:2560]: The last argument here should actually be the OS thread ID.  This allows users of the event to match the lock up with the thread that is holding it in a trace.  The memory address of the lock won't allow for this.

**Review by brianrob — COMMENTED** [review]
Looking good.  A couple of small things.

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: I didn't see any other paths that create sync blocks

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: I commented on this in https://github.com/dotnet/runtime/pull/72627#issuecomment-1224119479. It looks like the holding thread pointer cannot be dereferenced here because there's not guarantee that it would be alive. The `ThreadRunning` event is emitted from a newly started thread and can be used to map the `Thread` pointer to the OS thread ID.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: For threads that are created from outside the runtime and then enter the runtime, I believe the `ThreadCreated` event can be used to do the same mapping.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2560]: I don't think we can depend on the `ThreadRunning` event because there is nothing that guarantees that this event is present in the trace.  If the thread already exists when the trace is started, then we have no way to correlate it back to the OS thread.  That said, this call is made in a function that requires cooperative mode.  Given this, even if the thread dies, the memory shouldn't be reclaimed until after the next GC when the managed object Finalizer executes, right?

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2246]: From what I can see, there are others.  Here's an example: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/object.cpp#L987

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: It looks like this path also goes through `GetSyncBlock()`, which is modified here, so it would also emit the event

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2246]: Right.  There are others too, so I don't think this is the right place to emit the event, as it will be potentially much more verbose (and duplicative) than we want.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: The `AwareLock` is created when the `SyncBlock` is created, the monitor is an instance field. Even if the object hasn't been used for locks yet, the `AwareLock` would be created at this point.

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: > I don't think we can depend on the ThreadRunning event because there is nothing that guarantees that this event is present in the trace. If the thread already exists when the trace is started, then we have no way to correlate it back to the OS thread.

That's also true for the new `LockCreated` event. It seems like the ordering of those events would matter, so having a rundown of thread and lock objects may not help much to identify the thread. Perhaps the thread object is stored as the holding thread because it is easily accessible and is the minimum work necessary to identify the holding thread. Storing the OS thread ID instead would involve an indirection, and OS thread IDs may also be reused at some point, I suspect this kind of issue would exist in multiple places.

> That said, this call is made in a function that requires cooperative mode. Given this, even if the thread dies, the memory shouldn't be reclaimed until after the next GC when the managed object Finalizer executes, right?

I'm not sure about that, coop GC mode would ensure that the managed thread object would be alive if accessible, but once the managed thread object is deemed unreachable and the finalizer runs, perhaps the native thread object can be deleted irrespective of coop GC mode. Maybe there are further constraints that I'm not aware of?

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: Perhaps something that would fill the gap is to have an event that is emitted upon starting to collect a trace (not rundown), which maps the native thread object to an OS thread ID. Would that be feasible?

An alternative may be to ensure that it is ok to dereference the holding thread field somehow when emitting the event.

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2560]: We could just check if the `m_holdingThread` is null before dereferencing it to avoid crashing. Pass 0 if it's null, the OS Thread ID instead.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: What's preventing the `Thread` object from being deleted after reading a non-null value for the field?

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2560]: How would it be an issue  ?
The thread id would be informational and if someone wants to use (stackwalking for example), they will have to check if the thread is dead or not.

Comment (jkotas) [reply src/coreclr/vm/syncblk.cpp:2560]: If the thread is dead, reading the ID from the Thread* can crash the process - the memory for Thread* can be unmapped.

Comment (gleocadie) [reply src/coreclr/vm/syncblk.cpp:2560]: Oh ok, I was thinking @kouvel was referring to an invalid OS Thread ID, which is not a problem per se.
But yeah, accessing unmapped memory would be bad.

Thanks @jkotas for the clarification.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2560]: Ordering of `LockCreated` and `Contention` is important - if you don't have the `LockCreated` event, then it's possible that if the ID is re-used, that you won't know it.  It doesn't make the data useless, but could cause someone to conflate the existence of two locks down to one.  At the same time, emitting the `LockCreated` event more often than we should (e.g. every time there is a call to `GetSyncBlock`) would yield the opposite problem - we could see many creation events for a single lock creation.

With regard to the OS thread ID.  You're right that these can be re-used.  This is addressed because the OS will tell us when a thread is created/destroyed.  Thus, if there is a re-use within a trace, we'd see this from the OS thread events.

I was trying to convince myself that the managed thread object would solve the liveness issue, but I think you're right that it does not, since it's possible for there to not be a managed `Thread` object.  A way to address this would be for the thread that acquires the lock to save its thread ID into the lock when tracing is on and clear it when the lock is relinquished.  What do you think of that?  There are cases where this wouldn't work (e.g. the thread acquires the lock and blocks indefinitely before the trace is started), but I'm not too concerned about this, and still like the idea of keeping the overhead low, such that we don't save and clear the OS thread ID when tracing is disabled.

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2246]: That makes sense, and I think it makes sense at that point to log the event.  I just don't think we should log the event each time the syncblock is acquired, since that could happen much more often than the first time that it is created.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: > I just don't think we should log the event each time the syncblock is acquired, since that could happen much more often than the first time that it is created.

I thought the sync block would be created once per object and no more? I figured the event would be emitted when the sync block is created for an object, and the sync block may be reused for different purposes thereafter. Sync block creation could be frequent, but probably less frequent than lock contention.

Comment (jkotas) [reply src/coreclr/vm/syncblk.cpp:2246]: > I thought the sync block would be created once per object and no more? I

That's only true for precious syncblocks. Syncblocks that are not marked as precious can be reclaimed and re-created multiple times.

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: > emitting the LockCreated event more often than we should (e.g. every time there is a call to GetSyncBlock)

Why would the `LockCreated` event be emitted every time `GetSyncBlock()` is called?

> With regard to the OS thread ID. You're right that these can be re-used. This is addressed because the OS will tell us when a thread is created/destroyed. Thus, if there is a re-use within a trace, we'd see this from the OS thread events.

Why would there not be a similar problem to what you mentioned before regarding missing events when starting a trace on an already-running process? For instance, the create/destroy OS events for threads may not be captured by the trace.

> A way to address this would be for the thread that acquires the lock to save its thread ID into the lock when tracing is on and clear it when the lock is relinquished. What do you think of that?

The information stored is probably minimal, in the sense what is stored is what is most easily accessible for correct behavior, a different implementation may store something else and that should be ok. As I mentioned before, storing the thread ID would involve an indirection. Changing the implementation for this particular reason is probably something that would need a lot more scrutiny, I don't see enough reason for it yet.

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: > That's only true for precious syncblocks. Syncblocks that are not marked as precious can be reclaimed and re-created multiple times.

Wouldn't that mean that each time a sync block is recreated for an object, that it may have a different pointer to the `AwareLock` and hence warrant emitting the event?

Comment (jkotas) [reply src/coreclr/vm/syncblk.cpp:2246]: Yes, that's correct.

I think that Brian's concern is that we will emit this event for all SyncBlocks even the ones that are not ever used for locks. SyncBlocks are created for number of different reasons (the name "syncblock" is misleading). For example, interop creates them a lot for COM objects and marshalled delegates.

Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2246]: Yes.  What do you think about having the `AwareLock` log that it was created lazily - on first acquisition?

File: src/coreclr/vm/syncblk.cpp
Comment (brianrob) [reply src/coreclr/vm/syncblk.cpp:2560]: > Why would the LockCreated event be emitted every time GetSyncBlock() is called?

You're right, it shouldn't.  I missed the fact that the cache lookup occurs at the beginning of this function, and returns early.

> Why would there not be a similar problem to what you mentioned before regarding missing events when starting a trace on an already-running process? For instance, the create/destroy OS events for threads may not be captured by the trace.

You technically don't need to see the thread creation event to know that it existed at the beginning of the trace, as long as you see samples or events on the thread.  What you do need is to have a thread creation event during the trace to know that the thread ID has been re-used.  That said, the OS provides thread/process rundown events at the beginning of the trace to tell profilers about the existence of all processes and traces.

> The information stored is probably minimal, in the sense what is stored is what is most easily accessible for correct behavior, a different implementation may store something else and that should be ok. As I mentioned before, storing the thread ID would involve an indirection. Changing the implementation for this particular reason is probably something that would need a lot more scrutiny, I don't see enough reason for it yet.

I'm not sure I follow.  Can you tell me a little bit more here?  Are you saying that you don't think having the OS Thread ID is worthwhile here?

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2246]: I see. Yes, it could probably be emitted lazily. Taking a step back, is the lock ID really necessary? The stack trace of the contention event would probably be sufficient to identify the lock. I figure the `LockCreated` event's stack trace would provide similar information. A benefit may be to have only a few `LockCreated` events and avoid taking stack traces for contention events, but at the cost of missing those created events in some cases. Is there any other benefit?

File: src/coreclr/vm/syncblk.cpp
Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: > That said, the OS provides thread/process rundown events at the beginning of the trace to tell profilers about the existence of all processes and traces.

Curious, does this work with circular buffers? More generally, is there some mechanism by which a trace taken with a circular buffer, where the buffer fills up and is overwritten, can still have an accurate snapshot of live objects through events at the beginning of the trace?

> Are you saying that you don't think having the OS Thread ID is worthwhile here?

No, mapping to a thread ID would probably be the most useful part. I don't think we should change or add to what is stored for the lock-holding thread. Perhaps taking the thread store lock could enable emitting the OS thread ID.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp:2560]: A thread that orphans a lock does not clear the holding thread pointer before its `Thread` object is destroyed, so acquiring the thread store lock still would not make it feasible to dereference the holding thread pointer. There seem to be a few other related issues and fixing them properly would likely be more involved. For now, we could also store the OS thread ID when the event keyword is enabled as you suggested and update it after the other issues are fixed.

Comment (kouvel) [reply src/coreclr/vm/syncblk.cpp]: Updated in the latest commit to store and send the OS thread ID, since I think it would be possible to continue doing so after the complete fixes. Still curious about the question in https://github.com/dotnet/runtime/pull/72627#discussion_r955313883, as I've run into that issue before and a solution to it may be useful elsewhere.

**Comment by kouvel** [comment]
Updated in the latest commit to store and send the OS thread ID. I think the main remaining question is https://github.com/dotnet/runtime/pull/72627#discussion_r955306785. The `LockCreated` event wouldn't help to identify the managed object used for locking, which I'm not sure is that useful, and I don't see much additional benefit in having a `LockID`.

**Comment by tommcdon** [comment]
@kouvel @brianrob do you have additional thoughts on a path forward for this PR?  Do we need to actually consume these changes in tooling to prove/disprove the usefulness for lock contention investigations?

**Comment by brianrob** [comment]
Sorry, this got lost in my inbox.  I think that the latest change @kouvel added is great.

As for the remaining question, I am not quite sure how to address this.  Let's say that we don't have a creation event.  We will never know how many locks we're dealing with.  For example, let's say that there is a type of object that gets processed, and some of the work that happens must happen under lock.  The processing is likely to happen with the same stack over and over, but there may be a particular instance or instances that have more contention than others just based on the data.  Given that sync blocks can be re-used, it would be quite difficult for us to know how many locks are in play here - we'd have the LockID that would help, but as those IDs are re-used over time, we would not be able to tell.  This is a fairly classic pattern that we see across the runtime - other examples of this type of thing include assembly loads being assigned address space, jitted code for collectible assemblies, and GC heap survival events to report object movement during compaction.

I think that if we don't have a LockID or LockCreated event, and just try to depend upon the stack, the value of the additional information decays fairly heavily because you can no longer determine lock lifetime or which stacks are acquiring which lock.

**Comment by kouvel** [comment]
Having the `AwareLock` pointer available as the `LockID` would enable determining how many and which of the `AwareLock`s are contending a lot, but it may difficult if not impossible to identify from a trace which managed objects the heavily contending locks are associated with. Perhaps the info could be used as a trigger for heap dump collection for further investigation or something like that for cases where more info than the stack trace is necessary.

**Comment by jkotas** [comment]
Should the lock created event have `associatedObjectID` parameter that can be used to identify the primary object instance that the lock is associated with?

**Comment by brianrob** [comment]
Just paging this back into my memory, so let me know if I miss something or state the opposite of what I was wanting earlier.  I think using the `AwareLock` pointer is fine to identify the lock.  The key is that the pointer can be re-used, and so we won't know if it's been re-used unless we emit the `LockCreated` event.

> Should the lock created event have associatedObjectID parameter that can be used to identify the primary object instance that the lock is associated with?

If this is easily available, I think it's a great idea.  For long-running processes, capturing a dump and a trace together could provide some great insight into contention patterns.

**Comment by kouvel** [comment]
I guess having both the `AwareLock` pointer and the managed object pointer could be more useful. Initially the managed object pointer was being sent for the `LockID`, having the `AwareLock` makes it easier to tell how many locks are contending heavily.

**Comment by brianrob** [comment]
The one thing we'll need to be aware of is that the objectID can change and so we may need to use something like the heap survival and movement events to make sure that we get the right object id.  That said, I don't think we should spec that out completely just yet.

**Comment by kouvel** [comment]
Updated and added PerfView changes in https://github.com/microsoft/perfview/pull/1731

**Comment by tommcdon** [comment]
Thank you @kouvel @brianrob @jkotas!  Are there any other changes needed to merge this PR?

**Review by jkotas — APPROVED** [review]
LGTM

**Pull request merged**
